smolyay/surrogate.py

```diff
@@ -0,0 +1,540 @@
+                    .T.reshape(-1, self.dimension))
+                if grid_points_indexes is None:
+                    grid_points_indexes = grid_points_indexes_
+                else:
+                    grid_points_indexes = numpy.concatenate(
+                        (grid_points_indexes, grid_points_indexes_), axis=0)
+        self._grids_indexes = grid_points_indexes.tolist()
+        self._grids_basis = (
+            numpy.array(points)[grid_points_indexes].tolist())
+
+
+def generate_compositions(value, num_parts, include_zero):
+    """Generate compositions of a value into num_parts parts.
+
+    The algorithm that is being used is NEXCOM and can be found in
+    "Combinatorial Algorithms For Computers and Calculators",
+    Second Edition, 1978.
+    Authors: ALBERT NIJENHUIS and HERBERT S. WILF.
+    https://doi.org/10.1016/C2013-0-11243-3
+
+    ``include_zero`` parameter determines whether the compositions
+    that contain zeros are parts of the output or not.
+
+    The first composition will be (``value``, 0, ..., 0)
+    (or (``value-num_parts``, 0, ..., 0) if ``include_zero`` is ``False``).
+    Next, the first component (index = 0) is dropped by 1, and next component
+    (index = 1) is incremented by 1. This goes on until index 0 reaches ``0``.
+    Once, generated all compositions, the  next component is incremented
+    by 1 and is fixed until all compositions are generated with the same
+    method. This goes on until the last component reaches the ``value``
+    (or ``value - num_parts`` if ``include_zero`` is ``False``).
+    It is important to note that if ``include_zero`` is ``False``,
+    all components will be incremented by 1.
+
+    Parameters
+    ----------
+    value: int
+        Value.
+    num_parts: int
+        Number of parts.
+    include_zero : bool
+        True if compositions contain zero, False otherwise.
+
+    Yields
+    ------
+    list
+        All possible compositions of the value into num_parts.
+
+    Raises
+    ------
+    ValueError
+        Number of parts cannot be greater than value if the desired output
+        does not include compositions containing zeroes.
+    """
+    if value < num_parts and include_zero is False:
+        raise ValueError(
+            "When include_zero is {}, num_parts cannot be greater"
+            " than the value"
+            .format(False))
+    value = value if include_zero else value - num_parts
+
+    # (A) first entry
+    r = [0]*num_parts
+    r[0] = value
+    t = value
+    h = 0
+    yield list(r) if include_zero else (numpy.array(r)+1).tolist()
+
+    # (D)
+    while r[num_parts-1] != value:
+        # (B)
+        if t != 1:
+            h = 0
+
+        # (C)
+        h += 1
+        t = r[h-1]
+        r[h-1] = 0
+        r[0] = t-1
+        r[h] += 1
+        yield list(r) if include_zero else (numpy.array(r)+1).tolist()
+
+########################################################################
```

Comment:
I think this is duplicated from above? If so, remove.

Reply:
Why do you need it here if it is also in `grid.py`? If the code is the same, it should only be in one spot.